### ISSUE TITLE
Agents acquiring much more work than what they should

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#9816]
**Impact of the bug**
WMAgent

**Describe the bug**
Since this PR has been put in place:
https://github.com/dmwm/WMCore/pull/9679

I noticed that some agents acquire many hundreds of thousands of jobs, as can be seen in our WMAgent monitoring dashboard:
https://monit-grafana.cern.ch/d/lhVKAhNik/cms-wmagent-monitoring?orgId=11&refresh=5m&from=1594128943235&to=1594301743236&var-agent_url=All&var-site=All&var-priority=All

There must be a flaw in the WorkQueueBackend `availableWork` method logic that slipped through.

**How to reproduce it**
Not really sure

**Expected behavior**
The agent should respect the site thresholds and only pull work down if slots are available. This also means that one single agent should not pull too much work at once (200 elements could result in 200k jobs though, just too much!).

Instead of oversubscribing a site, we could also reduce the remaining jobs from other site slots, e.g.:
WQE has CERN and FNAL in the whitelist, 1k jobs expected.
if CERN has only 500 slots available,, then we would count 500 slots against CERN and 500 slots against FNAL. Instead of simply counting 1000 slots against CERN and moving on.

To be discussed: I have started considering partitioning the site resources as well. In the current schema, one single agent can potentially fill up all the site resources, i.e. if CERN has 40k slots, any single agent in the pool sees all 40k slots. Which means, they acquire much more work than they should, potentially creating a backlog in a single agent while others could have work for CERN as well...

**Additional context and error message**
none

**Comment by aperezca** [comment]
Hi Alan, Please excuse my ignorance, but why are we trying to tune job requests based on current used / idle slots? 

I understand that these are workloads already intended to run at a given site (or a list of sites), so they will just be eventually picked up by the agents and pushed to the queues. I mean, the destination of the jobs is already decided by Unified/WM, this is just the "materialization" of this decision, right?
If so, what is the downside of pulling as much work as readily available for each and every site? Yes, we would be loading the schedds and agents to higher levels, but also the glideinWMS FE would keep a higher pressure on the sites, so we can in principle also gain access to more resources, even opportunistically. 

In summary, By tuning workload down (e.g. stop submission if idle cpu = 0), we are just letting the FE know that no more glideins are needed!
Can you please explain the rationale on this?

**Comment by amaltaro** [comment]
Hi Antonio, sorry for the very belated reply. Here are my replies:

> Hi Alan, Please excuse my ignorance, but why are we trying to tune job requests based on current used / idle slots?

Given that any single agent in the pool have the ability to use all the resources for any given site, we need to find a way where we are sure that the load can be properly distributed among different agents (instead of leaving work queued in the agents local queue).

> I mean, the destination of the jobs is already decided by Unified/WM, this is just the "materialization" of this decision, right?

Jobs have a whitelist, but the data location of the chunk of work can restrict the job location to a subset of the original SiteWhitelist.

> If so, what is the downside of pulling as much work as readily available for each and every site?

Because there are WMAgent and HTCondor limits. For instance, HTCondor is set to 100k jobs in the queue. If we submit as many as we can, it could mean that higher priority work can't get submitted. Agents (and schedds) have other limitations, so it does not help to have one agent with 200k jobs, while others only have 20k. It's all about balance and prioritization.

**Comment by amaltaro** [comment]
From submit6 - running the latest WMAgent release - another side effect of this work prioritization and multi agent queue (local workqueue and WMBS):
```
2020-08-10 11:20:04,504:140190537946880:INFO:JobSubmitterPoller:Skipping cache update to be submitted. (91881 job in cache)
2020-08-10 11:20:05,283:140190537946880:INFO:JobSubmitterPoller:Submitter reached limit of submit slots for this cycle: 1000
2020-08-10 11:20:05,284:140190537946880:INFO:JobSubmitterPoller:Site submission report ...
2020-08-10 11:20:05,284:140190537946880:INFO:JobSubmitterPoller:    T1_IT_CNAF : {"Processing": {"NoTaskPendingSlot": 48914, "submitted": 18}}
2020-08-10 11:20:05,285:140190537946880:INFO:JobSubmitterPoller:    T3_US_PSC : {"Production": {"NoTaskPendingSlot": 83}}
2020-08-10 11:20:05,285:140190537946880:INFO:JobSubmitterPoller:    T1_RU_JINR : {"Processing": {"submitted": 25}}
2020-08-10 11:20:05,285:140190537946880:INFO:JobSubmitterPoller:    T2_CH_CERN_HLT : {"Processing": {"submitted": 816}}
2020-08-10 11:20:05,285:140190537946880:INFO:JobSubmitterPoller:    T2_DE_DESY : {"Processing": {"NoTaskPendingSlot": 52, "submitted": 62}}
2020-08-10 11:20:05,285:140190537946880:INFO:JobSubmitterPoller:    T1_ES_PIC : {"Processing": {"NoPendingSlot": 114, "submitted": 27}}
2020-08-10 11:20:05,285:140190537946880:INFO:JobSubmitterPoller:    T1_US_FNAL : {"Processing": {"submitted": 52}}
2020-08-10 11:20:05,286:140190537946880:INFO:JobSubmitterPoller:    T3_US_SDSC : {"Production": {"NoTaskPendingSlot": 83}}
2020-08-10 11:20:05,286:140190537946880:INFO:JobSubmitterPoller:    T3_US_NERSC : {"Production": {"NoTaskPendingSlot": 83}}
2020-08-10 11:20:05,286:140190537946880:INFO:JobSubmitterPoller:    T3_US_TACC : {"Production": {"NoPendingSlot": 83}}
2020-08-10 11:20:05,286:140190537946880:INFO:JobSubmitterPoller:Priority submission report ...
2020-08-10 11:20:05,286:140190537946880:INFO:JobSubmitterPoller:    300000.0 : {"Production": {"Total": 83}}
2020-08-10 11:20:05,286:140190537946880:INFO:JobSubmitterPoller:    220400.0 : {"Processing": {"Total": 48932, "submitted": 18}}
2020-08-10 11:20:05,286:140190537946880:INFO:JobSubmitterPoller:    110080.0 : {"Processing": {"Total": 982, "submitted": 982}}
2020-08-10 11:20:05,287:140190537946880:INFO:JobSubmitterPoller:Have 33 packages to submit.
2020-08-10 11:20:05,287:140190537946880:INFO:JobSubmitterPoller:Have 1000 jobs to submit.
```

a single agent should never have ~50k jobs created for one single site (T1_IT_CNAF), especially when those resources can only run 10k jobs...:
```
[cmsdataops@cmsgwms-submit6 current]$ $manage execute-agent wmagent-resource-control --site-name=T1_IT_CNAF -p
Executing wmagent-resource-control --site-name=T1_IT_CNAF -p ...
Thresholds and current status for all sites:

T1_IT_CNAF - 229 running, 1386 pending, 10001 running slots total, 1500 pending slots total, Site is Normal:
```

In short, work acquisition needs to be re-though and reworked. Basically the same thing that I've been saying since 2018...

**Comment by aperezca** [comment]
Hi Alan,
I was asking about workload acquisition and queueing in WM layer as a whole, not specifically at the "per agent" level, (which I did not get from the context of the issue). Having said that, and considering your arguments of multiple limitations for every agent, I think this looks like a "planned distribution" vs "competition" problem, and I tend to lean more on the "free market" approach on this one... In that sense, I'd also reiterate myself and add that having a sufficiently high number of agents pulling work simultaneously in competition with each other would probably help in achieving the dynamic balance between agents that WM requires. I'd definitely try that, besides setting some obvious thresholds per agent. 

I know your objections towards running the system with the larger amount of machines possible, but I think we keep running into issues affecting global production levels where the bottleneck can ultimately be attributed to the low number of agents (or at least, their impact would be reduced if we had more agents).

Let me stress again that what I consider the best option for scheduling as a whole is to try to move as much workload and as soon as possible into the schedd queues, then let condor do the rest. That's the best way to maximize pressure on the pool and benefit from every opportunistic cycle we can find.

**Comment by amaltaro** [comment]
> Let me stress again that what I consider the best option for scheduling as a whole is to try to move as much workload and as soon as possible into the schedd queues, then let condor do the rest. That's the best way to maximize pressure on the pool and benefit from every opportunistic cycle we can find.

I agree that this guarantees a high pressure on the pool, thus a better usage of the hardware resources. However, it has a pretty negative impact on the person power, given that there would be more jobs in condor (including lower priority jobs), those would timeout for being in condor for too long, and after a few retries, they would permanently fail (thus, adding more load on the P&R team to recover from failures).
In addition to that, having a "free market" model where you submit as many jobs as you can to condor, leaves no room for higher priority work that needs to start running as soon as possible (given that this model would likely run at the schedd limitations).

**Comment by aperezca** [comment]
Some follow-up questions:

"there would be more jobs in condor (including lower priority jobs), those would timeout for being in condor for too long" why would they timeout? do jobs "expire" if they are in a schedd queue and don't get to run/complete in X amount of time? even if so, we could still tune this value of course...

"a "free market" model where you submit as many jobs as you can to condor, leaves no room for higher priority work that needs to start running as soon as possible (given that this model would likely run at the schedd limitations)"
What do you mean? I think we can agree that once jobs are in the condor queues, highest priority jobs, compatible with the available resources, would always run first, so I don't see the problem in having also a high number of low priority requests there as well. Or are you suggesting that the limitation would come from the step of injecting new high-prio requests into the schedd queues? you think we would saturate the schedd queues even when running a large set of agents/schedds?

**Comment by amaltaro** [comment]
> "there would be more jobs in condor (including lower priority jobs), those would timeout for being in condor for too long" why would they timeout? do jobs "expire" if they are in a schedd queue and don't get to run/complete in X amount of time? even if so, we could still tune this value of course...

Yes, there is a component watching jobs in Pending and Running state, if they go beyond an agent-wide configuration value, the component will remove them. Right now, jobs can't be pending for more than 5 days (see it as a protection to ensure we have the proper jobs in the condor queue).

> What do you mean? I think we can agree that once jobs are in the condor queues, highest priority jobs, compatible with the available resources, would always run first,

Agreed!

> so I don't see the problem in having also a high number of low priority requests there as well. Or are you suggesting that the limitation would come from the step of injecting new high-prio requests into the schedd queues? you think we would saturate the schedd queues even when running a large set of agents/schedds?

The problem I'm trying to avoid is to keep jobs in condor when they have no chance at all to run. That's why I dislike this  idea - submit all you can to condor as soon as possible - it adds extra load in all the WM system, it increases the operational burden on multiple layers as well (WM, PnR, SI), and we would only benefit during those period of times that we have issues somewhere in the production system.

Having said it all, I think one of the ways forward would be to implement a better resource management within WMCore (global would be the best), and make sure our set of agents always have work pending for sites providing resources (provided there are workflows requesting those resources).

**Comment by amaltaro** [comment]
Just an update to problem originally reported here. There must be a very special scenario that allows agents to acquire more GQEs than what they were supposed to. I've been looking at the agents for the last couple of weeks and they are pretty stable - and even - regarding the amount of work pulled down.

One issue that needs to be investigated and addressed though - related to this very same workqueue code - concerns the lack of jobs for some sites, even though there are workqueue elements available for them (e.g. CSCS). This happens when GQE has a lower priority, and the agent never manages to reach those elements when pulling down up to X (=200) elements from global workqueue.
Having said that, we need to make sure that sites are not left empty, even if the only work for them comes from very low priority workflows.

**Comment by amaltaro** [comment]
> One issue that needs to be investigated and addressed though - related to this very same workqueue code - concerns the lack of jobs for some sites, even though there are workqueue elements available for them (e.g. CSCS). This happens when GQE has a lower priority, and the agent never manages to reach those elements when pulling down up to X (=200) elements from global workqueue.

This issue should have been resolved by: https://github.com/dmwm/WMCore/pull/10072

We are left with the original issue reported in this issue, which might actually come from the taskchain blow up factor, still to be investigated...

**Comment by amaltaro** [comment]
I still see agents here and there acquiring tons of work, not very often though. Right now, we have vocms0250 with almost 200k jobs in the queue. Problem is, it's not clear to me whether it's really a problem, or just high priority work coming in and sites having a very large capacity of slots (given that there is only 2 agents up and running at the moment).

I think the best way to deal with this case is going to be through this issue: https://github.com/dmwm/WMCore/issues/10301

**Comment by amaltaro** [comment]
Pushing it to Q2.
